src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1311,7 +1313,23 @@ protected override void Dispose(bool finalize)
                 _projectionBufferFactoryService.ProjectionBufferCreated -= AddTextBufferCloneServiceToBuffer;
                 FileWatchedReferenceFactory.ReferenceChanged -= RefreshMetadataReferencesForFile;
 
-                _deferredJoinableTasks.Join();
+                // https://github.com/Microsoft/vs-threading/blob/master/doc/cookbook_vs.md#how-to-write-a-fire-and-forget-method-responsibly
+                _disposalTokenSource.Cancel();
```

Comment:
Why the copy of this pattern here? I would have assumed this just used the implementation you just added in IThreadingContext.

Reply:
The code in `IThreadingContext` is only usable for MEF parts that aren't going to be disposed at some other point in time. It effectively means the call to `IThreadingContext.Dispose` is disposing of these parts.

Reply:
For this that's not really a concern -- the VisualStudioWorkspace is a MEF part that's also being disposed at the same time.